### PR TITLE
cGuard v3.3 Protocol support

### PR DESCRIPTION
**Pull request by mpak59rus**

### CODE REVIEW

File: src/org/traccar/protocol/CguardProtocol.java [hunk 1]
```diff
@@ -0,0 +1,51 @@
+/*
+ * Copyright 2015 Anton Tananaev (anton.tananaev@gmail.com)
```
Comment (drecchia) [inline src/org/traccar/protocol/CguardProtocol.java]: Please change year to the current year.

File: src/org/traccar/protocol/CguardProtocolDecoder.java [hunk 2]
```diff
@@ -0,0 +1,136 @@
+/*
+ * Copyright 2012 - 2015 Anton Tananaev (anton.tananaev@gmail.com)
```
Comment (drecchia) [inline src/org/traccar/protocol/CguardProtocolDecoder.java]: Its a new class, copyright should be 2016.

File: src/org/traccar/protocol/CguardProtocol.java [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+    public CguardProtocol() {
+        super("cguard");
+    }
+
+    @Override
+    public void initTrackerServers(List<TrackerServer> serverList) {
+        serverList.add(new TrackerServer(new ServerBootstrap(), this.getName()) {
+            @Override
+            protected void addSpecificHandlers(ChannelPipeline pipeline) {
+                //pipeline.addLast("frameDecoder", new LengthFieldBasedFrameDecoder(256, 2, 1, 2, 0));
```
Comment (tananaev) [inline src/org/traccar/protocol/CguardProtocol.java]: TCP server needs frame decoder.

File: src/org/traccar/protocol/CguardProtocol.java [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+                pipeline.addLast("objectDecoder", new CguardProtocolDecoder(CguardProtocol.this));
+            }
+        });
+        serverList.add(new TrackerServer(new ConnectionlessBootstrap(), this.getName()) {
+            @Override
+            protected void addSpecificHandlers(ChannelPipeline pipeline) {
+               /* pipeline.addLast("stringEncoder", new StringEncoder());
```
Comment (tananaev) [inline src/org/traccar/protocol/CguardProtocol.java]: Do we really need UDP? Also, please remove comments.

File: src/org/traccar/protocol/CguardProtocolDecoder.java [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+    @Override
+    protected Object decode(Channel channel, SocketAddress remoteAddress, Object msg) throws Exception {
+
+        ChannelBuffer buf = (ChannelBuffer) msg;
+
+        String bufString = buf.toString(Charset.defaultCharset());
+        String[] bufStringRows = bufString.split("\n");
+        String dataString = bufStringRows[0];
+
+        if(dataString.startsWith(VERSION_DATA)) {
```
Comment (tananaev) [inline src/org/traccar/protocol/CguardProtocolDecoder.java]: Please fix formatting everywhere. There should be a space after "if" keyword.

File: test/org/traccar/protocol/CguardDecoderTest.java [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+
+import org.junit.Test;
+import org.traccar.ProtocolTest;
+
+public class CguardDecoderTest extends ProtocolTest {
+
+    @Test
+    public void testDecode() throws Exception {
+
+        CguardProtocolDecoder decoder = new CguardProtocolDecoder(new CguardProtocol());
```
Comment (tananaev) [inline test/org/traccar/protocol/CguardDecoderTest.java]: Please provide unit tests with a couple of cases.

**Comment by tananaev** [comment]
There are quite a few comment. Also, please fix merge issues.

**Comment by tananaev** [comment]
I have reimplemented cGuard decoder myself. The code is in 19b17695867e00861c719292aad7ae48f33a8eeb. It would be nice if you could test it.

**Pull request closed without merging**